src/Features/Core/Portable/UnusedReferences/IUnusedReferencesService.cs

```diff
@@ -0,0 +1,31 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.UnusedReferences
+{
+    internal interface IUnusedReferencesService
+    {
+        /// <summary>
+        /// Determines unused references from this compilation.
+        /// </summary>
+        Task<ImmutableArray<UnusedReference>> GetUnusedReferencesAsync(
+            Project project,
+            Compilation compilation,
+            CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Updates references by removing, adding, or marking references as TreatAsUsed
+        /// for this project.
+        /// </summary>
+        /// <returns>Updated project</returns>
+        Task<Project> UpdateReferencesAsync(
+            Project project,
+            ImmutableArray<ReferenceUpdate> referenceUpdates,
+            CancellationToken cancellationToken);
```

Comment:
Not clear to me yet what this is needed for

Reply:
The Unused Reference UI would invoke this to apply the changes the user has choosen. Removing direct references, adding transitive references, or updating a TrestAsUsed reference attribute.

Reply:
a few more questions:

1. "Removing direct references".  If they have a Project instance, why can't htey just do the removal themselve and give us back the new Project instance?
2. "or updating a TrestAsUsed reference attribute."  I feel like we want a bit on MetadataREference to say that it is TreatAsUsed.  With that bit all hte UI has to do is fork teh Project instance and replace the MetadataReferences with teh appropriate forked versions with that bit set.
3. "adding transitive references,".  Can you expand on this.  Why is that part of an 'unused reference ui'?

Reply:
@CyrusNajmabadi Thanks for your questions. My goal for this service is that it provides the high-level Project and Package references that are unused and allows those high-level references to be updated.

1. So, the UI will invoke this method with the list of changes that the user requested. This method will then work with the project system to ensure that the Project and Package references are updated accordingly.
2. A single Project or Package references is likely composed of many MetadataReferences. I am also unsure how an updated metadata reference results in an updated project file. Unless you are saying each of these updates could be encoded as changes in a project file.
3. It is possible that a no types from a directly referenced Project or Package are used in the compilation, but types from a tansitive reference it brings along are used. We could provide a means to remove the unused direct reference and directly add the transitive reference in these cases.

Reply:
> It is possible that a no types from a directly referenced Project or Package are used in the compilation, but types from a tansitive reference it brings along are used. We could provide a means to remove the unused direct reference and directly add the transitive reference in these cases.

i don't think that's possible... can you give an example.  i believe the compielr needs all references to be direct to understand anything.

Reply:
I think "meta packages" (Microsoft.Net.Compilers.Toolset are the simplest example where nothing is being used from the direct reference, but transitive references are being used.